### PR TITLE
Issue #1245 Change endpoints processing in cp-edge

### PR DESCRIPTION
**Pull request by Wedds**
This PR is related to issue #1245 
It changes the way of endpoints' processing in cp-edge service:
- system endpoints are configured via JSON file
- system endpoints can be applied to 'non-interactive' tools

### CODE REVIEW

File: deploy/docker/cp-edge/sync-routes.py [hunk 1]
```diff
@@ -245,24 +246,41 @@ def search_custom_domain_cert(domain):
...
+        for endpoint in tool_endpoints:
+                endpoint_obj = json.loads(endpoint)
+                tool_endpoints_dict[endpoint_obj['name']] = endpoint_obj['nginx']['port']
         if run_details and "pipelineRunParameters" in run_details:
                 # Get a list of endpoints from SYSTEM_ENDPOINTS which match the run's parameters (param name and a value)
-                system_endpoints_matched = [SYSTEM_ENDPOINTS[x["name"]] for x in run_details["pipelineRunParameters"] 
-                                                if x["name"] in system_endpoints_params 
+                system_endpoints_matched = [SYSTEM_ENDPOINTS[x["name"]] for x in run_details["pipelineRunParameters"]
+                                                if x["name"] in system_endpoints_params
+                                                   and (x["friendly_name"] not in tool_endpoints_dict
```
Comment (ekazachkova) [inline deploy/docker/cp-edge/sync-routes.py]: The `PipelineRunParameter` object hasn't `friendly_name` field. So, this line may lead to `KeyError`.

**Review by ekazachkova — CHANGES_REQUESTED** [review]

File: deploy/docker/cp-edge/sync-routes.py [hunk 2]
```diff
@@ -245,24 +246,41 @@ def search_custom_domain_cert(domain):
...
+SYSTEM_ENDPOINTS = read_system_endpoints()
+
 def append_system_endpoints(tool_endpoints, run_details):
         if not tool_endpoints:
                 tool_endpoints = []
         system_endpoints_params = SYSTEM_ENDPOINTS.keys()
+        tool_endpoints_dict = {}
+        for endpoint in tool_endpoints:
+                endpoint_obj = json.loads(endpoint)
+                tool_endpoints_dict[endpoint_obj['name']] = endpoint_obj['nginx']['port']
```
Comment (ekazachkova) [inline deploy/docker/cp-edge/sync-routes.py]: May the `endpoint_obj['nginx']` returns `None`? ( [check for that case](https://github.com/epam/cloud-pipeline/pull/1267/files#diff-cb34f51c1b0b798757a8b597c39e6f9fR354) )

**Review by ekazachkova — CHANGES_REQUESTED** [review]

File: deploy/docker/cp-edge/sync-routes.py (hunk 1 above)
Comment (Wedds) [reply deploy/docker/cp-edge/sync-routes.py]: Done, `x` in this and the following line is replaced with `SYSTEM_ENDPOINTS[x["name"]]`

File: deploy/docker/cp-edge/sync-routes.py (hunk 2 above)
Comment (Wedds) [reply deploy/docker/cp-edge/sync-routes.py]: Done, checking if both `name` and `nginx` are presented in the `endpoint_obj` after conversion from JSON

File: deploy/docker/cp-edge/sync-routes.py [hunk 3]
```diff
@@ -273,16 +286,48 @@ def append_system_endpoints(tool_endpoints, run_details):
...
-                        tool_endpoint = { "nginx": { "port": system_endpoint["endpoint"] }, "isDefault": "false" }
+                        tool_endpoint = { "nginx": { "port": system_endpoint["endpoint"] }}
+                        system_endpoint_port = system_endpoint["endpoint"]
+                        system_endpoint_name = None
                         if "friendly_name" in system_endpoint:
                                 tool_endpoint["name"] = system_endpoint["friendly_name"]
+                                system_endpoint_name = system_endpoint["friendly_name"]
                         if "endpoint_num" in system_endpoint and system_endpoint["endpoint_num"]:
                                 tool_endpoint["endpoint_num"] = system_endpoint["endpoint_num"]
+                        modified_tool_endpoints, modified_count, isDefaultEndpoint = \
```
Comment (ekazachkova) [inline deploy/docker/cp-edge/sync-routes.py]: Lets continue to use naming convention `lower_case_with_underscores` in this script

File: deploy/docker/cp-edge/sync-routes.py [hunk 4]
```diff
@@ -273,16 +286,48 @@ def append_system_endpoints(tool_endpoints, run_details):
...
+                                modified_endpoints_count += modified_count
                         tool_endpoints.append(json.dumps(tool_endpoint))
-        return tool_endpoints 
+        return tool_endpoints, modified_endpoints_count
+
+
+def remove_from_tool_endpoints_if_fully_matches(endpoint_name, endpoint_port, tool_endpoints):
+        tools_endpoints_wo_matches = []
+        endpoints_modified = 0
+        isModifiedDefault = False
```
Comment (ekazachkova) [inline deploy/docker/cp-edge/sync-routes.py]: Lets continue to use naming convention `lower_case_with_underscores` in this script

File: deploy/docker/cp-edge/sync-routes.py [hunk 5]
```diff
@@ -273,16 +286,48 @@ def append_system_endpoints(tool_endpoints, run_details):
...
+                        if modified_count != 0:
+                                tool_endpoints = modified_tool_endpoints
+                                modified_endpoints_count += modified_count
                         tool_endpoints.append(json.dumps(tool_endpoint))
-        return tool_endpoints 
+        return tool_endpoints, modified_endpoints_count
+
+
+def remove_from_tool_endpoints_if_fully_matches(endpoint_name, endpoint_port, tool_endpoints):
+        tools_endpoints_wo_matches = []
```
Comment (ekazachkova) [inline deploy/docker/cp-edge/sync-routes.py]: Is it correct field name?

File: deploy/docker/cp-edge/sync-routes.py [hunk 6]
```diff
@@ -412,17 +463,35 @@ def get_service_list(pod_id, pod_run_id, pod_ip):
...
+def load_pods_for_runs_with_endpoints():
+        pods_with_endpoints = []
+        all_pipeline_pods = Pod.objects(kube_api).filter(selector={'type': 'pipeline'})\
+                                                 .filter(field_selector={"status.phase": "Running"})
+        for pod in all_pipeline_pods.response['items']:
+                labels = pod['metadata']['labels']
+                if 'job-type' in labels and labels['job-type'] == 'Service':
+                        pods_with_endpoints.append(pod)
+                        continue
+                pipeline_env_parameters = pod['spec']['containers'][0]['env']
```
Comment (ekazachkova) [inline deploy/docker/cp-edge/sync-routes.py]: Should we add check 'spec' and 'containers' for existence and `pod['spec']['containers'][0]` for not null?

File: deploy/docker/cp-edge/sync-routes.py (hunk 3 above)
Comment (ekazachkova) [inline deploy/docker/cp-edge/sync-routes.py]: According to this naming, it looks like `modified_count` = `len(modified_tool_endpoints)` but this is not true. Maybe we should make more understandable names?

**Review by ekazachkova — CHANGES_REQUESTED** [review]

File: deploy/docker/cp-edge/sync-routes.py (hunk 3 above)
Comment (Wedds) [reply deploy/docker/cp-edge/sync-routes.py]: Done

File: deploy/docker/cp-edge/sync-routes.py (hunk 4 above)
Comment (Wedds) [reply deploy/docker/cp-edge/sync-routes.py]: Done

File: deploy/docker/cp-edge/sync-routes.py (hunk 5 above)
Comment (Wedds) [reply deploy/docker/cp-edge/sync-routes.py]: Renamed to more suitable `non_matching_tool_endpoints`

File: deploy/docker/cp-edge/sync-routes.py (hunk 6 above)
Comment (Wedds) [reply deploy/docker/cp-edge/sync-routes.py]: Done, perform additional checks, that response structure is as expected

File: deploy/docker/cp-edge/sync-routes.py (hunk 3 above)
Comment (Wedds) [reply deploy/docker/cp-edge/sync-routes.py]: More convenient prefixes are used for those values

**Review by ekazachkova — APPROVED** [review]

**Pull request merged**